zerver/lib/message.py

```diff
@@ -1212,3 +1215,35 @@ def get_recent_private_conversations(user_profile: UserProfile) -> Dict[int, Dic
         rec['user_ids'].sort()
 
     return recipient_map
+
+def wildcard_mention_allowed(sender: UserProfile, stream: Stream) -> bool:
+    realm = sender.realm
+
+    # If there are fewer than Realm.WILDCARD_MENTION_THRESHOLD, we
+    # allow sending.  In the future, we may want to make this behavior
+    # a default, and also just allow explicitly setting whether this
+    # applies to a stream as an override.
+    if num_subscribers_for_stream_id(stream.id) <= Realm.WILDCARD_MENTION_THRESHOLD:
+        return True
+
+    if realm.wildcard_mention_policy == Realm.WILDCARD_MENTION_POLICY_NOBODY:
+        raise False
+
+    if realm.wildcard_mention_policy == Realm.WILDCARD_MENTION_POLICY_EVERYONE:
+        return True
+
+    if realm.wildcard_mention_policy == Realm.WILDCARD_MENTION_POLICY_ADMINS:
+        return sender.is_realm_admin
+
+    if realm.wildcard_mention_policy == Realm.WILDCARD_MENTION_POLICY_STREAM_ADMINS:
+        # TODO: Change this when we implement stream administrators
+        return sender.is_realm_admin
+
+    if realm.wildcard_mention_policy == Realm.WILDCARD_MENTION_POLICY_FULL_MEMBERS:
+        return not sender.is_new_member and not sender.is_guest
+
+    if realm.wildcard_mention_policy == Realm.WILDCARD_MENTION_POLICY_MEMBERS:
+        return not sender.is_guest
+
+    raise AssertionError("Invalid wildcard mention policy")
```

Comment:
I rewrote this function to have a cleaner style and support the full set of policy options we'd like to have.